### ISSUE TITLE
ActiveSupport's isolation does not work with popen

### ISSUE THREAD

**Issue by headius** [rails/rails#19777]
ActiveSupport isolation, used for isolating test runs from each other, normally tries to use fork to launch those subprocesses. If fork is unavailable or the NO_FORK env var is set, it will use popen.

It appears that the popen logic is not being tested, because it doesn't work. Running with Ruby 2.2 and NO_FORK=1:

```
[] ~/projects/rails/activesupport $ NO_FORK=1 rake test
/Users/headius/.rvm/rubies/ruby-2.2.0/bin/ruby -w -I"lib:test"  "/Users/headius/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/rake/rake_test_loader.rb" "test/**/*_test.rb" 
Run options: --seed 8212

# Running:

..........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................F......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................../Users/headius/projects/rails/activesupport/lib/active_support/testing/isolation.rb:82:in `popen': no implicit conversion of Symbol into String (TypeError)
	from /Users/headius/projects/rails/activesupport/lib/active_support/testing/isolation.rb:82:in `block in run_in_isolation'
	from /Users/headius/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/tempfile.rb:319:in `open'
	from /Users/headius/projects/rails/activesupport/lib/active_support/testing/isolation.rb:70:in `run_in_isolation'
	from /Users/headius/projects/rails/activesupport/lib/active_support/testing/isolation.rb:28:in `run'
	from /Users/headius/.rvm/gems/ruby-2.2.0/gems/minitest-5.3.3/lib/minitest.rb:759:in `run_one_method'
	from /Users/headius/.rvm/gems/ruby-2.2.0/gems/minitest-5.3.3/lib/minitest/parallel.rb:14:in `block (2 levels) in initialize'
rake aborted!
Command failed with status (1): [ruby -w -I"lib:test"  "/Users/headius/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/rake/rake_test_loader.rb" "test/**/*_test.rb" ]
/Users/headius/.rvm/gems/ruby-2.2.0/bin/ruby_executable_hooks:15:in `eval'
/Users/headius/.rvm/gems/ruby-2.2.0/bin/ruby_executable_hooks:15:in `<main>'
Tasks: TOP => test
(See full trace by running task with --trace)
```

This unfortunately prevents JRuby from running many of the Rails suites *at all* since we do not support forking.

I attempted to fix this by making the hash in question (a literal hash in the popen isolation logic) use String keys, but then something else blew up and I have not investigated that yet:

```
[] ~/projects/rails/activesupport $ NO_FORK=1 rake test
/Users/headius/.rvm/rubies/ruby-2.2.0/bin/ruby -w -I"lib:test"  "/Users/headius/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/rake/rake_test_loader.rb" "test/**/*_test.rb" 
Run options: --seed 43875

# Running:

........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................F........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................../Users/headius/projects/rails/activesupport/lib/active_support/testing/isolation.rb:82: warning: openpath: pathname too long (ignored)
	File "/Users/headius/.rvm/rubies/ruby-2.2.0/bin/ruby -I"/Users/headius/.rvm/gems/ruby-2.2.0/gems/w3c_valid"...

/Users/headius/projects/rails/activesupport/lib/active_support/testing/isolation.rb:82: warning: openpath: pathname too long (ignored)
	File "/Users/headius/.rvm/rubies/ruby-2.2.0/bin/ruby -I"/Users/headius/.rvm/gems/ruby-2.2.0/gems/w3c_valid"...

/Users/headius/projects/rails/activesupport/lib/active_support/testing/isolation.rb:82:in `popen': No such file or directory - /Users/headius/.rvm/rubies/ruby-2.2.0/bin/ruby -I"/Users/headius/.rvm/gems/ruby-2.2.0/gems/w3c_validators-1.2/lib" -I"/Users/headius/.rvm/gems/ruby-2.2.0/gems/uglifier-2.7.0/lib" -I"/Users/headius/.rvm/gems/ruby-2.2.0/gems/turbolinks-2.5.3/lib" -I"/Users/headius/.rvm/gems/ruby-2.2.0/gems/sucker_punch-1.3.2/lib" -I"/Users/headius/.rvm/gems/ruby-2.2.0/gems/stackprof-0.2.7/lib" -I"/Users/headius/.rvm/gems/ruby-2.2.0/extensions/x86_64-darwin-14/2.2.0/stackprof-0.2.7" -I"/Users/headius/.rvm/gems/ruby-2.2.0/gems/sqlite3-1.3.10/lib" -I"/Users/headius/.rvm/gems/ruby-2.2.0/extensions/x86_64-darwin-14/2.2.0/sqlite3-1.3.10" -I"/Users/headius/.rvm/gems/ruby-2.2.0/gems/sneakers-0.1.1.pre/lib" -I"/Users/headius/.rvm/gems/ruby-2.2.0/gems/thread-0.1.5/lib" -I"/Users/headius/.rvm/gems/ruby-2.2.0/gems/sidekiq-3.3.2/lib" -I"/Users/headius/.rvm/gems/ruby-2.2.0/gems/serverengine-1.5.10/lib" -I"/Users/headius/.rvm/gems/ruby-2.2.0/gems/sigdump-0.2.2/lib" 
...
```

**Comment by gazay** [comment]
Here are a few problems

* First of all – `IO.popen` works different in mri and jruby (look at square brackets and output of echo command), and in jruby 9.0.0.0-pre1 it throws error any way:

```ruby
# ruby 2.2.2p95 (2015-04-13 revision 50295) [x86_64-darwin13]

irb(main):001:0> IO.popen([{'PATH' => 'something'}, 'echo $PATH']).read
Errno::ENOENT: No such file or directory - echo $PATH...
irb(main):002:0> IO.popen({'PATH' => 'something'}, 'echo $PATH').read
=> "something\n"

# jruby 1.7.19

irb(main):001:0> IO.popen([{'PATH' => 'something'}, 'echo $PATH']).read
=> "$PATH\n"
irb(main):002:0> IO.popen({'PATH' => 'something'}, 'echo $PATH').read
ArgumentError: illegal access mode echo $PATH...

# jruby 9.0.0.0 pre1
irb(main):001:0> IO.popen([{'PATH' => 'some'}, 'echo $PATH'])
Errno::ENOENT: No such file or directory - echo $PATH...
[ERROR] Failed to disable echo
java.io.IOException: Cannot run program "sh": error=2, No such file or directory

irb(main):001:0> IO.popen({'PATH' => 'some'}, 'echo $PATH').read
=> "some\n"
[ERROR] Failed to disable echo
java.io.IOException: Cannot run program "sh": error=2, No such file or directory...
```

* In mri `IO.popen` makes strange unwrap of path mask if we send command like `…rake_test_loader.rb test/**/*_test.rb` – it just unwraps it in `*_test.rb` files from INSIDE of test/** directories but not in `test/*_test.rb` – so test from `test/autoload_test.rb` can not be found. That's why for mri it should be wrapped like that: `...rake_test_loader.rb 'test/**/*_test.rb' ...`.

* Activesupport specs from master not working with any version on jruby

I’ve changed code of ActiveSupport::Testing::Isolation::Subprocess that it runs on mri: https://github.com/gazay/rails/commit/61103ac7b0801201533e2c853957da4d80d81b51

I can make a PR with this change, but I’m not sure, because purpose of this code is to work with jruby and ruby on windows (which I can’t currently check) and in it’s current condition it is not working in mri and jruby. So making it work only for mri would have not so many sense. What do you think @rafaelfranca?

My thoughts that we should make `IO.popen` in jruby compatible with mri behavior.

**Comment by rafaelfranca** [comment]
@gazay lets fix first for MRI and lets work to get JRuby fixed later. But lets keep this issue open until we have both working.

**Comment by headius** [comment]
IO.popen in JRuby 9k.pre2 should match MRI for all scenarios. Since this change is only needed for Rails 5/master (right?) and only JRuby 9k supports Ruby 2.0+ APIs, I think it's ok to rely on 9k.pre2 behavior.

```
irb(main):002:0> IO.popen([{'PATH' => 'something'}, 'echo $PATH']).read
Errno::ENOENT: No such file or directory - echo $PATH
	from org/jruby/RubyIO.java:3829:in `popen'
	from (irb):2:in `<eval>'
	from org/jruby/RubyKernel.java:1005:in `eval'
	from org/jruby/RubyKernel.java:1315:in `loop'
	from org/jruby/RubyKernel.java:1125:in `catch'
	from org/jruby/RubyKernel.java:1125:in `catch'
	from /Users/headius/projects/jruby/bin/jirb:13:in `<top>'
irb(main):003:0> IO.popen({'PATH' => 'something'}, 'echo $PATH').read
=> "something\n"
```

I believe 1.7.20 can also be fixed to do the right thing in the second case (the case we expect to work).

**Comment by headius** [comment]
I have made various fixes to popen for JRuby 1.7.20 that should handle these forms better.

**Comment by gazay** [comment]
My fixes in ActiveSupport::Testing::Isolation::Subprocess (https://github.com/rails/rails/issues/19819) works with current master of jruby

**Comment by headius** [comment]
I saw this got closed automatically...but @gazay claimed his fixes in #19819 work with JRuby. I think this could be marked as fixed for Rails 5 rather than just closed as stale.

**Comment by rafaelfranca** [comment]
Indeed I forgot to close the issue when I merged that PR. Thank you for pointing it out!
